File: android/app/src/main/kotlin/ai/arcvanta/arcvanta/vision/ModelContract.kt
package ai.arcvanta.arcvanta.vision

/**
 * The Kotlin half of `vision/contract/model_contract.json`.
 *
 * Nothing here is chosen locally. Every value is read off the contract, and
 * the graphs are checked against it at load time so a mismatched export fails
 * loudly on startup instead of quietly producing landmarks in the wrong place.
 */
object ModelContract {
    const val VERSION = 1

    const val DETECTOR_ASSET = "arcvanta_rtmdet_m_640.onnx"
    const val POSE_ASSET = "arcvanta_rtmpose_m_256x192.onnx"

    const val DETECTOR_INPUT = "images"
    const val DETECTOR_SIZE = 640
    const val DETECTOR_PAD_VALUE = 114

    /**
     * BGR order, because RTMDet sets `bgr_to_rgb=False` and was trained on
     * channels the way OpenCV loads them. The pose model below is the
     * opposite, so the two cannot share a channel order and each preprocessing
     * loop has to say which one it is writing.
     */
    val DETECTOR_MEAN = floatArrayOf(103.53f, 116.28f, 123.675f)
    val DETECTOR_STD = floatArrayOf(57.375f, 57.12f, 58.395f)

    const val POSE_INPUT = "input"
    const val POSE_WIDTH = 192
    const val POSE_HEIGHT = 256
    const val POSE_PADDING_FACTOR = 1.25f
    const val SIMCC_SPLIT_RATIO = 2.0f
    const val KEYPOINT_COUNT = 17

    /** RGB order, matching `bgr_to_rgb=True` in the mmpose config. */
    val POSE_MEAN = floatArrayOf(123.675f, 116.28f, 103.53f)
    val POSE_STD = floatArrayOf(58.395f, 57.12f, 57.375f)

    const val CLASS_PERSON = 0
    const val CLASS_BALL = 1
    const val CLASS_RIM = 2
    const val CLASS_BACKBOARD = 3

    /**
     * Below this a detection is not worth forwarding. Kept low because the
     * detector was trained on broadcast footage and scores lower on close-range
     * phone frames. The calibration solver grades the quality separately.
     */
    const val SCORE_THRESHOLD = 0.15f
}

File: ios/Runner/Vision/ModelContract.swift
import Foundation

/// The Swift half of `vision/contract/model_contract.json`.
///
/// Nothing here is chosen locally. Every value is read off the contract, and
/// the graphs are checked against it at load time so a mismatched export fails
/// loudly on startup instead of quietly producing landmarks in the wrong place.
enum ModelContract {
    static let version = 1

    static let detectorAsset = "arcvanta_rtmdet_m_640.onnx"
    static let poseAsset = "arcvanta_rtmpose_m_256x192.onnx"

    static let detectorInput = "images"
    static let detectorSize = 640
    static let detectorPadValue: Float = 114

    /// BGR order, because RTMDet sets `bgr_to_rgb=False` and was trained on
    /// channels the way the frame arrives. The pose constants below are the
    /// opposite, so the two preprocessing loops cannot be shared.
    static let detectorMean: [Float] = [103.53, 116.28, 123.675]
    static let detectorStd: [Float] = [57.375, 57.12, 58.395]

    static let poseInput = "input"
    static let poseWidth = 192
    static let poseHeight = 256
    static let posePaddingFactor: Float = 1.25
    static let simccSplitRatio: Float = 2.0
    static let keypointCount = 17

    static let poseMean: [Float] = [123.675, 116.28, 103.53]
    static let poseStd: [Float] = [58.395, 57.12, 57.375]

    static let classPerson = 0
    static let classBall = 1
    static let classRim = 2
    static let classBackboard = 3

    /// Below this a detection is not worth forwarding.
    static let scoreThreshold: Float = 0.35
}
